rfd/0056-buildbox-image-handling.md

```diff
@@ -0,0 +1,90 @@
+
+ 3. Changes to the buildbox image must be reviewable and auditable.
+
+ 4. Changes to the buildbox should be opt-in, in the sense that changes to a 
+    buildbox should not be *forced* on a consumer - the consumer should deliberately 
+    reference the updated buildbox image.
+
+ 5. Using the buildbox images (at least in public-facing CI) must *not* require 
+    the use of privileged containers (e.g. require Docker-In-Docker), as this 
+    is a security threat.
+
+ 6. Creating an updated buildbox image once changes have been approved should 
+    be automatic (or at least a single push-button operation)
+
+ 7. Where multiple build environments are required, they should all be 
+    produced by the same process.
+
+### The current process
+
+Currently the buildbox images are constructed by Drone in response to a commits
+being pushed onto `gravitational/teleport:master`. The drone pipeline constructs 
+updated buildbox images from the latest source in `build.assets/Dockerfile` (and 
+it's siblings), tags them with the version of `go` installed on the image and 
+then pushed them to the repository at `quay.io`. These builds are then used for CI.
+
+For _releases_, the release pipelines in Drone construct a buildbox image at 
+_build time_ from the same source in `build.assets`, and then execute the build 
+inside it via Docker-in-Docker
+
+This process meets several of the goals set above:
+
+ * The buildbox for a release are constructed from source at build time, so the 
+   current process (somewhat - I'll explain why below) meets Goal #1.
+ * Changes to the buildbox are incorporated by the PR/code review process, so 
+   Goal #3 is met
+ * Images are automatically produced for both CI and Release, so Goal #4 is met
+ * Images for all of the different build targets are created in the same way, so
+   Goal #7 is met
+
+On the other hand:
+
+ * We do no validation of the components installed onto the buildbox image, so 
+   the every image created (including at build time for releases) is vulnerable
+   to downloaded artifacts being subverted.
+ * We tag _every_ revision of the buildbox with the current go version, meaning
+   that we cannot uniquely identify any revision of the buildbox image.
+ * Using the version of Go as the image tag means we have no mechanism to 
+   independently upgrade other software in the image, or identify which images 
+   contain what versions of, for example, Rust.
+ * Updating the buildbox image automatically forces and CI jobs using the 
+   buildbox to accept the new changes. This has recently caused issues where 
+   updating the buildbox `go1.17.2` image to build Teleport 9 broke CI for 
+   Teleport 8, as there is no way to differentiate the images they require.
+
+### Suggested changes
+
+* Independently versioned buildboxes, so we can deal with things like Go and 
+  Rust being updated independently
+* Buildbox source extracted from Teleport repo into it's own space
```

Comment:
I don't think this is a win -- but I'd want to dig into the idea further.

My gut reaction is: if buildbox source lives in a seperate repo, that either means:

1) we use something like dependencies or submodules, and every single commit before now becomes two commits. One to make the change in the hypothetical buildbox repo, and a second to update the dependency/submodule.
2) We have changes outside of the teleport repo (e.g. a single commit to the hypothetical buildbox repo) that are auto-include in teleport.  This will result in things breaking without associated changes in the teleport repo, which is quite confusing to debug.

In short, I think buildbox source should probably remain in the teleport repo, for both convenience and determinism.

Reply:
If it's not a clear win, it should probably stay in the teleport repo. My reasoning is

1. The whole `build.assets` directory is a murky part of the teleport repo that I don't think people understand, so removing it from the teleport repo and having appropriate READMEs & so forth wouldbring it out into the open
2. An scenario that requires pre-built images for CI is _going_ to require a 2-step shuffle, as you can't build the image internally and then run a build on it without requiring DinD or the like.
3. If CI and release are using strongly-versioned (as opposed to floating) tags (which I would argue they should be), changes to the buildbox are opt-in to the main repo. 
4. Personal inertia: This is how I've done it in the past, so it feels natural to me. I accept that this rationale may not scale :-)

Reply:
> Any scenario that requires pre-built images for CI or release is going to require a 2-step shuffle, as you can't build the image internally and then run a build on it without requiring DinD or the like.

I see.  I think the DinD concern was the element I was not considering.  You're right that I don't know of a way to have JIT buildboxes without a way to build docker images. DinD + a stronger isolation mechanism (e.g. short lived single purpose VMs or gVisor) woudl do the trick but  that is way out of scope here.

Now that I understand the full context, I'm ok with this.  I don't love it due to the busy work concern mentioned above, but nor do I see a better option in the short term.